### DISCUSSION TITLE
Help with custom Lens<TextureAtlasSprite>

### DISCUSSION THREAD

**Discussion by njelly** [djeedai/bevy_tweening#51] [Q&A] [answered]
I have successfully implemented a custom lens for a Transform component (this "orbits" a transform around a point at a distance):
```
struct TranslateAroundLens {
    start_rot: Quat,
    end_rot: Quat,
    distance: f32,
    point: Vec3,
}

impl Lens<Transform> for TranslateAroundLens {
    fn lerp(self: &mut TranslateAroundLens, target: &mut Transform, ratio: f32) {
        let rot = self.start_rot.lerp(self.end_rot, ratio);
        target.translation = rot.mul_vec3(Vec3::new(self.distance, 0., 0.)) + self.point;
    }
}
```

Now I want to create Lens<TextureAtlasSprite> and have the tween run on the same entity, but I can't get it to work at all (should be identical to SpriteColorLens?). It simply doesn't seem to run, although I can see that the Animator<TextureAtlasSprite> is definitely getting added to my entity:

```
struct TextureAtlasSpriteColorLens {
    start: Color,
    end: Color,
}

impl Lens<TextureAtlasSprite> for TextureAtlasSpriteColorLens {
    fn lerp(self: &mut TextureAtlasSpriteColorLens, target: &mut TextureAtlasSprite, ratio: f32) {
        let start: Vec4 = self.start.into();
        let end: Vec4 = self.end.into();
        let value = start.lerp(end, ratio);
        target.color = value.into();
        println!("this never prints!");
    }
}
```

Here is the pastebin for my full module:
https://pastebin.com/raw/Rmhsmqiv

**Comment by njelly** [comment, marked as answer]
So I realize now that I need to add a `component_animator_system::<TextureAtlasSprite>` to my app, however this exists in the `plugin` module, which is private, so I'm not sure how this is intended to be implemented? Seems like this should **_not_** be private, but the solution for now is to just copy+paste that function definition into my code base and use it there.

**Reply by djeedai** [reply]
Sorry yes you need that system which is public : https://docs.rs/bevy_tweening/latest/bevy_tweening/fn.component_animator_system.html

https://github.com/djeedai/bevy_tweening/blob/2c9e4a2ae0d4ae972147f1505063549c30f538f3/src/lib.rs#L165

**Reply by njelly** [reply]
So I get this error when I do that, which is why I'm wondering if this is a mistake or if I'm missing something else? Either way I solved it by just copying the definition of component_animator_system into my own codebase and adding my systems in my own plugin

```
module `plugin` is private
private module rustc[E0603](https://doc.rust-lang.org/error-index.html#E0603)
```
